src/model_theory/basic.lean

```diff
@@ -0,0 +1,329 @@
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+
+## References
+For the Flypitch project:
+- [J. Han, F. van Doorn, *A formal proof of the independence of the continuum hypothesis*]
+[flypitch_cpp]
+- [J. Han, F. van Doorn, *A formalization of forcing and the unprovability of
+the continuum hypothesis*][flypitch_itp]
+
+-/
+
+universes l u
+
+variables {α : Type u} {β : Type u}
+
+namespace first_order
+
+/-- A first-order language consists of a type of functions of every natural-number arity and a
+  type of relations of every natural-number arity. -/
+structure language : Type (l+1) :=
+(functions : ℕ → Type l) (relations : ℕ → Type l)
+
+namespace language
+
+/-- The empty language has no symbols. -/
+def empty : language := ⟨λ _, pempty, λ _, pempty⟩
+
+instance : inhabited language := ⟨empty⟩
+
+/-- The type of constants in a given language. -/
+@[nolint has_inhabited_instance] def const (L : language.{l}) : Type l := L.functions 0
+
+variable (L : language.{l})
+
+/-- A language is relational when it has no function symbols. -/
+class is_relational : Prop :=
+(empty_functions : ∀ n, L.functions n → false)
+
+/-- A language is algebraic when it has no relation symbols. -/
+class is_algebraic : Prop :=
+(empty_relations : ∀ n, L.relations n → false)
+
+variable {L}
+
+instance is_relational_of_empty_functions {symb : ℕ → Type l} : is_relational ⟨λ _, pempty, symb⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+instance is_algebraic_of_empty_relations {symb : ℕ → Type l}  : is_algebraic ⟨symb, λ _, pempty⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+instance is_relational_empty : is_relational (empty) := language.is_relational_of_empty_functions
+instance is_algebraic_empty : is_algebraic (empty) := language.is_algebraic_of_empty_relations
+
+variables (L) (M : Type u)
+
+/-- A first-order structure on a type `M` consists of interpretations of all the symbols in a given
+  language. Each function of arity `n` is interpreted as a function sending tuples of length `n`
+  (modeled as `(fin n → M)`) to `M`, and a relation of arity `n` is a function from tuples of length
+  `n` to `Prop`. -/
+class Structure :=
+(fun_map : ∀{n}, L.functions n → (fin n → M) → M)
+(rel_map : ∀{n}, L.relations n → (fin n → M) → Prop)
+
+variables (N : Type u) [L.Structure M] [L.Structure N]
+
+open first_order.language.Structure
+
+/-- A homomorphism between first-order structures is a function that commutes with the
+  interpretations of functions and maps tuples in one structure where a given relation is true to
+  tuples in the second structure where that relation is still true. -/
+protected structure hom :=
+(to_fun : M → N)
+(map_fun' : ∀{n} (f : L.functions n) x, to_fun (fun_map f x) = fun_map f (to_fun ∘ x) . obviously)
+(map_rel' : ∀{n} (r : L.relations n) x, rel_map r x → rel_map r (to_fun ∘ x) . obviously)
+
+notation A ` →[`:25 L `] ` B := L.hom A B
```

Comment:
This clashes with the `mul_action_hom` notation (unless it is a special Unicode bracket).

Reply:
I've asked Zulip for advice on unique notation for this.

Reply:
Could you just use localized? I think all but the most fundamental notations should be hidden under `open_locale`.

Reply:
I've localized my notation in `first_order`, but I haven't changed it otherwise.